src/fairseq2/cluster.py

```diff
@@ -182,3 +187,108 @@ def supports_current_cluster(self) -> bool:
     @override
     def supported_cluster(self) -> str:
         return "none"
+
+
+@ray.remote(num_cpus=0)
+class RayCoordinator:
+    NAME = "RAY_FAIRSEQ2_COORDINATOR_NAME"
+    LEADER_MAX_RETRIES = 30
+    LEADER_RETRY_INTERVAL = 1.0
+
+    def __init__(
+        self,
+        job_id: int,
+        world_size: int,
+    ):
+        self.job_id = job_id
+        self.worker_info: Dict[int, Dict[str, Any]] = {}
+        self.leader_port = None
+        self.ready_workers = 0
+        self.world_size = world_size
+
+    def register_worker(
+        self, hostname: str, rank: int, free_port: int | None
+    ) -> Dict[str, Any]:
+        """Register a worker with its placement group ID and GPU ID"""
+        self.ready_workers += 1
+        info = {
+            "hostname": hostname,
+            "rank": rank,
+            "ready_workers": self.ready_workers,
+            "world_size": self.world_size,
+            "leader_port": free_port,
+        }
+        self.worker_info[rank] = info
+        return info
+
+    def get_leader_info(self) -> Dict[str, Any] | None:
+        if self.ready_workers == self.world_size:
+            return self.worker_info[0]
+        else:
+            return None
+
+
+@final
+class RayClusterHandler(ClusterHandler):
+    _env: MutableMapping[str, str]
+
+    def __init__(self, env: MutableMapping[str, str]) -> None:
+        self._env = env
+
+    @override
+    def set_torch_distributed_variables(self) -> None:
+        env = self._env
+
+        rank_str = env.get("RANK")
```

Comment:
You can (ideally should) use the `get_rank`, `get_local_rank`, `get_world_size`, and `get_local_world_size` utility functions defined here: https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/utils/env.py#L85

They also raise an appropriate exception that causes the process to terminate gracefully.

Reply:
thanks, changed to use utility functions.